### PR TITLE
Fixed #340

### LINKED ISSUE #340: Stop queue threads on changing data sets
In attempt to stop server crashes:

- Stop any training threads that feed the training queue before starting validation
- Stop any validation threads that feed the validation queue before starting training
- Stop any validation/training threads that feed the validation/training queue before starting testing

In addition stop any internal queue threads in similar situations

### PR DESCRIPTION
**Pull request by kdavis-mozilla**
Properly stopped all Python Threads on all possible transitions:

* training-to-validation
* validation-to-training
* validation-to-testing
* training-to-testing

using a [tf.train.Coordinator](https://www.tensorflow.org/api_docs/python/train/coordinator_and_queuerunner#Coordinator).

The newly documented means of employing a [tf.train.Coordinator](https://www.tensorflow.org/api_docs/python/train/coordinator_and_queuerunner#Coordinator) to stop Python Threads internal or external to [QueueRunner](https://www.tensorflow.org/how_tos/threading_and_queues/#queuerunner)'s was employed.

### CODE REVIEW

File: DeepSpeech.ipynb [hunk 1]
```diff
@@ -1444,13 +1444,16 @@
...
     "        # Start queue runner threads\n",
-    "        tf.train.start_queue_runners(sess=session)\n",
+    "        managed_threads = tf.train.start_queue_runners(sess=session, coord=coord)\n",
     "\n",
     "        # Start importer's queue threads\n",
-    "        execution_context[1].start_queue_threads(session)\n",
+    "        managed_threads = managed_threads + execution_context[1].start_queue_threads(session, coord)\n",
     "\n",
-    "    return session"
+    "    return session, coord, managed_threads"
```
Comment (tilmankamp) [inline DeepSpeech.ipynb]: One could think of bundling `session, coord, managed_threads` into another tuple. Just a thought - it's ok like that.

File: util/importers/LDC97S62.py [hunk 2]
```diff
@@ -51,6 +52,7 @@ def __init__(self, txt_files, thread_count, batch_size, numcep, numcontext):
                                                   dtypes = [tf.float32, tf.int32, tf.int32, tf.int32],
                                                   capacity = 2 * self._get_device_count() * batch_size)
         self._enqueue_op = self._example_queue.enqueue([self._x, self._x_length, self._y, self._y_length])
+        self._close_op = self._example_queue.close(cancel_pending_enqueues=True)
```
Comment (tilmankamp) [inline util/importers/LDC97S62.py]: Much more reliable and graceful.

File: util/importers/LDC97S62.py [hunk 3]
```diff
@@ -94,7 +102,7 @@ def _populate_batch_queue(self, session):
                     self._x_length: source_len,
                     self._y: target,
                     self._y_length: target_len})
-            except (RuntimeError, tf.errors.CancelledError):
+            except Exception as e:
```
Comment (tilmankamp) [inline util/importers/LDC97S62.py]: Why is this required? I once tried to make it as specific as possible. Now it captures everything and by that it'd make debugging much harder.

File: util/importers/LDC97S62.py [hunk 4]
```diff
@@ -82,6 +88,8 @@ def _create_files_circular_list(self):
 
     def _populate_batch_queue(self, session):
         for txt_file, wav_file in self._files_circular_list:
+            if self._coord.should_stop():
```
Comment (tilmankamp) [inline util/importers/LDC97S62.py]: I like this one.

File: util/importers/fisher.py [hunk 5]
```diff
@@ -91,7 +100,7 @@ def _populate_batch_queue(self, session):
                     self._x_length: source_len,
                     self._y: target,
                     self._y_length: target_len})
-            except (RuntimeError, tf.errors.CancelledError):
+            except Exception as e:
```
Comment (tilmankamp) [inline util/importers/fisher.py]: Why is this required? I once tried to make it as specific as possible. Now it captures everything and by that it'd make debugging much harder.

File: util/importers/ldc93s1.py [hunk 6]
```diff
@@ -74,14 +80,14 @@ def _compute_source_target(self):
...
-        while True:
+        while not self._coord.should_stop():
             try:
                 session.run(self._enqueue_op, feed_dict={
                     self._x: source,
                     self._x_length: source_len,
                     self._y: target,
                     self._y_length: target_len})
-            except (RuntimeError, tf.errors.CancelledError):
+            except Exception as e:
```
Comment (tilmankamp) [inline util/importers/ldc93s1.py]: Why is this required? I once tried to make it as specific as possible. Now it captures everything and by that it'd make debugging much harder.

File: util/importers/librivox.py [hunk 7]
```diff
@@ -97,7 +105,7 @@ def _populate_batch_queue(self, session):
                     self._x_length: source_len,
                     self._y: target,
                     self._y_length: target_len})
-            except (RuntimeError, tf.errors.CancelledError):
+            except Exception as e:
```
Comment (tilmankamp) [inline util/importers/librivox.py]: Why is this required? I once tried to make it as specific as possible. Now it captures everything and by that it'd make debugging much harder.

File: util/importers/ted.py [hunk 8]
```diff
@@ -105,7 +113,7 @@ def _populate_batch_queue(self, session):
                     self._x_length: source_len,
                     self._y: target,
                     self._y_length: target_len})
-            except (RuntimeError, tf.errors.CancelledError):
+            except Exception as e:
```
Comment (tilmankamp) [inline util/importers/ted.py]: Why is this required? I once tried to make it as specific as possible. Now it captures everything and by that it'd make debugging much harder.

File: util/importers/LDC97S62.py (hunk 3 above)
Comment (kdavis-mozilla) [inline util/importers/LDC97S62.py]: Yeah it's to general. I can switch it.

However, it should not be switched it back to how it was. The old code was also in error too.

We should **not** have caught `RuntimeError`. `RuntimeError` is thrown when we close a session, then call `run()` on it. We should never close a session then call `run()` on it as then the session is then in invalid state. Catching `RuntimeError` here was hiding part of the cause of the crashing bug.

Catching `tf.errors.CancelledError` seems reasonable, but I am not sure if this is the only thing we need to catch.

I can switch it, then re-run the traning-validation transition 50 more times to see if a problem occurs.

File: util/importers/fisher.py (hunk 5 above)
Comment (kdavis-mozilla) [inline util/importers/fisher.py]: Yeah it's to general. I can switch it.

However, it should not be switched it back to how it was. The old code was also in error too.

We should **not** have caught `RuntimeError`. `RuntimeError` is thrown when we close a session, then call `run()` on it. We should never close a session then call `run()` on it as then the session is then in invalid state. Catching `RuntimeError` here was hiding part of the cause of the crashing bug.

Catching `tf.errors.CancelledError` seems reasonable, but I am not sure if this is the only thing we need to catch.

I can switch it, then re-run the traning-validation transition 50 more times to see if a problem occurs.

File: util/importers/ldc93s1.py (hunk 6 above)
Comment (kdavis-mozilla) [inline util/importers/ldc93s1.py]: Yeah it's to general. I can switch it.

However, it should not be switched it back to how it was. The old code was also in error too.

We should **not** have caught `RuntimeError`. `RuntimeError` is thrown when we close a session, then call `run()` on it. We should never close a session then call `run()` on it as then the session is then in invalid state. Catching `RuntimeError` here was hiding part of the cause of the crashing bug.

Catching `tf.errors.CancelledError` seems reasonable, but I am not sure if this is the only thing we need to catch.

I can switch it, then re-run the traning-validation transition 50 more times to see if a problem occurs.

File: util/importers/librivox.py (hunk 7 above)
Comment (kdavis-mozilla) [inline util/importers/librivox.py]: Yeah it's to general. I can switch it.

However, it should not be switched it back to how it was. The old code was also in error too.

We should **not** have caught `RuntimeError`. `RuntimeError` is thrown when we close a session, then call `run()` on it. We should never close a session then call `run()` on it as then the session is then in invalid state. Catching `RuntimeError` here was hiding part of the cause of the crashing bug.

Catching `tf.errors.CancelledError` seems reasonable, but I am not sure if this is the only thing we need to catch.

I can switch it, then re-run the traning-validation transition 50 more times to see if a problem occurs.

File: util/importers/ted.py (hunk 8 above)
Comment (kdavis-mozilla) [inline util/importers/ted.py]: Yeah it's to general. I can switch it.

However, it should not be switched it back to how it was. The old code was also in error too.

We should **not** have caught `RuntimeError`. `RuntimeError` is thrown when we close a session, then call `run()` on it. We should never close a session then call `run()` on it as then the session is then in invalid state. Catching `RuntimeError` here was hiding part of the cause of the crashing bug.

Catching `tf.errors.CancelledError` seems reasonable, but I am not sure if this is the only thing we need to catch.

I can switch it, then re-run the traning-validation transition 50 more times to see if a problem occurs.

**Comment by kdavis-mozilla** [comment]
@tilmankamp Made the exception catching changes. Currently testing them on the server.
